README.md
# git-flow (CJS Edition)

A collection of Git extensions to provide high-level repository operations
for Vincent Driessen's [branching model](http://nvie.com/git-model "original
blog post"). This fork adds functionality not added to the original branch.

## Why another git-flow fork
The last commit to [gitflow-avh](https://github.com/petervanderdoes/gitflow-avh)
was on May 23, 2019 and has been archived on Jun 19, 2023. Since 2019 there have
been a number of issues opened that have not be resolved. This fork will address
those outstanding issues and open PR's along with continueing to maintain the
git-flow branching model.

## Getting started

For the best introduction to get started with `igit`, please read Jeff
Kreeftmeijer's blog post:

[Using git-flow to automate your git branching workflow](https://jeffkreeftmeijer.com/git-flow/)

Or have a look at one of these screen casts:

* [How to use a scalable Git branching model called git-flow](https://www.youtube.com/watch?v=-8DgENtFWSU) (by Build a Module)
* [A short introduction to git-flow](https://vimeo.com/16018419) (by Mark Derricutt)
* [On the path with git-flow](https://vimeo.com/37408017) (by Dave Bock)

A quick cheatsheet was made by Daniel Kummer:

[https://danielkummer.github.io/git-flow-cheatsheet/](https://danielkummer.github.io/git-flow-cheatsheet/)

## Installing git-flow

See the Wiki for up-to-date [Installation Instructions](https://github.com/CJ-Systems/gitflow-cjs/wiki/Installation).


## Integration with your shell

For those who use the [Bash](https://www.gnu.org/software/bash/) or [ZSH](https://www.zsh.org/)
shell, you can use my [fork of git-flow-completion](https://github.com/petervanderdoes/git-flow-completion)
which includes several additions for git-flow (AVH Edition), or you can use the
original [git-flow-completion](https://github.com/bobthecow/git-flow-completion)
project by [bobthecow](https://github.com/bobthecow). Both offer tab-completion
for git-flow subcommands and branch names with my fork including tab-completion
for the commands not found in the original git-flow.


## FAQ

* See the [FAQ](https://github.com/CJ-Systems/gitflow-cjs/wiki/FAQ) section
of the project Wiki.
* Version Numbering Scheme.  
Starting with version 1.0, the project uses the following scheme:
\<MAJOR\>.\<MINOR\>.\<REVISION\>\
* CJS is the acronym of "CJ Systems"

## Please help out

This project is under constant development. Feedback and suggestions are very
welcome and I encourage you to use the [Issues
list](https://github.com/CJ-Systems/gitflow-cjs/issues) on Github to provide that
feedback.

Feel free to fork this repository and to commit your additions. For a list of
all contributors, please see the [AUTHORS](AUTHORS) file.

Any questions, tips, or general discussion can be posted to the Google group:
[http://groups.google.com/group/gitflow-users](https://groups.google.com/g/gitflow-users)
This is the original group set up to support the nvie branch, but I am monitoring
the list as well for any questions related to my version.
When you do post a question on the list please indicate which version you are,
using the complete version number.

## Contributing

If submiting a new pull request addressing an already open issue with gitflow-avh please link the relevant issue in the description. For any new issues please see below

### Quick Start for new issues

* Please fork and clone a local copy of [gitflow-cjs](https://github.com/CJ-Systems/gitflow-cjs).
* Create a seperate issue branch based off develop.
* Commit commit you fix to the local branch.
* Please update your local copy with the latest devlop branch of [gitflow-cjs](https://github.com/CJ-Systems/gitflow-cjs)
* Rebase develop onto your local branch.
* Push your fix to your fork.
* When ready to submit a pull request.

### How to submit a pull request

For any new PRs releated to gitflow-cjs you can use on of the keywords from [Linking a pull request to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to automatically close the releated issue.

## License terms

git-flow is published under the FreeBSD License, see the
[LICENSE](LICENSE) file. Although the FreeBSD License does not require you to
share any modifications you make to the source code, you are very much
encouraged and invited to contribute back your modifications to the community,
preferably in a Github fork, of course.


## igit usage

### Initialization

To initialize a new repo with the basic branch structure, use:

    igit init [-d]

This will then interactively prompt you with some questions on which branches
you would like to use as development and production branches, and how you
would like your prefixes be named. You may simply press Return on any of
those questions to accept the (sane) default suggestions.

The ``-d`` flag will accept all defaults.

![Screencast igit init](https://i.imgur.com/lFQbY5V.gif)

### Creating feature/release/hotfix/support branches

* To list/start/finish/delete feature branches, use:

```shell
igit feature
igit feature start <name> [<base>]
igit feature finish <name>
igit feature delete <name>
```

  For feature branches, the `<base>` arg must be a branch, when omitted it defaults to the develop branch.

* To push/pull a feature branch to the remote repository, use:

```shell
igit feature publish <name>
igit feature track <name>
```

* To list/start/finish/delete release branches, use:

```shell
igit release
igit release start <release> [<base>]
igit release finish <release>
igit release delete <release>
```

  For release branches, the `<base>` arg must be a branch, when omitted it defaults to the develop branch.

* To list/start/finish/delete hotfix branches, use:

```shell
igit hotfix
igit hotfix start <release> [<base>]
igit hotfix finish <release>
igit hotfix delete <release>
```

  For hotfix branches, the `<base>` arg must be a branch, when omitted it defaults to the production branch.

* To list/start support branches, use:

```shell
igit support
igit support start <release> <base>
```

  For support branches, the `<base>` arg must be a branch, when omitted it defaults to the production branch.

### Share features with others

You can easily publish a feature you are working on. The reason can be to allow other programmers to work on it or to access it from another machine. The publish/track feature of gitflow simplify the creation of a remote branch and its tracking.

When you want to publish a feature just use:
```shell
igit feature publish <name>
```

or, if you already are into the `feature/<name>` branch, just issue:
```shell
igit feature publish
```

Now if you execute `git branch -avv` you will see that your branch `feature/<name>` tracks `[origin/feature/<name>]`. To track the same remote branch in another clone of the same repository use:
```shell
igit feature track <name>
```

This will create a local feature `feature/<name>` that tracks the same remote branch as the original one, that is `origin/feature/<name>`.

When one developer (depending on your work flow) finishes working on the feature he or she can issue `igit feature finish <name>` and this will automatically delete the remote branch. All other developers shall then run:
```shell
    igit feature delete <name>
```

to get rid of the local feature that tracks a remote branch that no more exist.

### Share hotfixes with others

You can publish an hotfix you are working on. The reason can be to allow other programmers to work on it or validate it or to access it from another machine.

When you want to publish an hotfix just use (as you did for features):
```shell
igit hotfix publish <name>
```

or, if you already are into the `hotfix/<name>` branch, just issue:
```shell
igit hotfix publish
```

Other developers can now update their repositories and checkout the hotfix:
```shell
git pull
git checkout hotfix/<name>
```
and eventually finish it:
```shell
igit hotfix finish
```


### Using Hooks and Filters

For a wide variety of commands hooks or filters can be called before and after
the command.  
The files should be placed in .git/hooks  
In the directory hooks you can find examples of all the hooks available.


## Showing your appreciation

Of course, the best way to show your appreciation for the git-flow tool itself
remains contributing to the community.  If you'd like to show your appreciation
in another way, however, consider donating through PayPal:

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/donate/?hosted_button_id=GU25NCHMVMT9U)



CHANGELOG.md
# Changelog

#### 2.2.1
* fix issue #82 remove color support for cherry-pick hotfix thank you [jrebmann](https://github.com/jrebmann) for reporting this

#### 2.2.0
* feat(init): add option to sign initial commit (#59)
* feat(git-flow): add support for environment credentials (#19)
* feat(git-flow): add support for busybox(#12)
* feat(feature): add feature release command (#64)
* feat(hotfix): add cherrypick option to hotfix finish (#73) 
* feat(hotfix): add  backmerge option to release branch (#71)
* fix(hotfix): igit hotfix finish -b still back-merges to develop (#66)
* fix(hotfix): igit hotfix track fails due to branch name check (#68)
* fix: igit log help message (#15)

#### 2.1.0
* feat: add create command to release allowing for a single step release prodution
* feat: add missing hook script filter-flow-release-finish-version
* fix: multi-line tag message when using -m
* docs: update readme with better description as to why this fork was created. Thank You [shaedrick](https://github.com/shaedrich)
**NOTE:** The format for a multi-line tag message is now "$(printf 'line of text with \n escape codes placed as needed for proper formating of the message')"

#### 2.0.1
* fix incorrect version identification along with updating source repository in gitflow-installer. Corrections pointed out by [bobstuart](https://github.com/bobstuart)

#### 2.0.0
* Bugfix/help text and reorder rename args Thank You [bmcdonnell](https://github.com/bmcdonnell)
* Cherry-Pick  Speed up "list" command execution #325 from gpongelli/gitflow-avh/tree/speedUpListCmd
* Updated readme links Thank You [JamesSkemp](https://github.com/JamesSkemp)
* Support gitflow-cjs being used as a plugin for zsh users via antigen. Thank You [matthewfranglen](https://github.com/matthewfranglen)


test.sh
GIT_URL=$(git config --get remote.origin.url)
access_token="1234567890"
authenticated_git_url="${GIT_URL/\:\/\//://$access_token@}"
echo $authenticated_git_url


contrib/igit-installer.sh
#!/bin/bash

# git-flow make-less installer for *nix systems, by Rick Osborne
# Based on the git-flow core Makefile:
# http://github.com/petervanderdoes/gitflow-avh/blob/master/Makefile

# Licensed under the same restrictions as git-flow:
# http://github.com/CJ-Systems/gitflow-cjs/blob/develop/LICENSE

# Updated for the fork at CJ-Systems

usage() {
	echo "Usage: [environment] igit.sh [install|uninstall]"
	echo "Environment:"
	echo "   PREFIX=$PREFIX"
	echo "   REPO_HOME=$REPO_HOME"
	echo "   REPO_NAME=$REPO_NAME"
	exit 1
}

# Does this need to be smarter for each host OS?
if [ -z "$PREFIX" ] ; then
	PREFIX="/usr/local"
fi

if [ -z "$REPO_NAME" ] ; then
	REPO_NAME="igit"
fi

if [ -z "$REPO_HOME" ] ; then
	REPO_HOME="https://github.com/ccfiel/gitflow-cjs.git"
fi

EXEC_PREFIX="$PREFIX"
BINDIR="$EXEC_PREFIX/bin"
DATAROOTDIR="$PREFIX/share"
DOCDIR="$DATAROOTDIR/doc/gitflow"

EXEC_FILES="igit"
SCRIPT_FILES="git-flow-init git-flow-feature git-flow-bugfix git-flow-hotfix git-flow-release git-flow-stage git-flow-support git-flow-version gitflow-common gitflow-shFlags git-flow-config"
HOOK_FILES="$REPO_NAME/hooks/*"


echo "### igit no-make installer ###"

case "$1" in
uninstall)
	echo "Uninstalling igit from $PREFIX"
	if [ -d "$BINDIR" ] ; then
		for script_file in $SCRIPT_FILES $EXEC_FILES ; do
			echo "rm -vf $BINDIR/$script_file"
			rm -vf "$BINDIR/$script_file"
		done
		rm -rf "$DOCDIR"
	else
		echo "The '$BINDIR' directory was not found."
	fi
	exit
	;;
help)
	usage
	exit
	;;
install)
	echo "Installing igit to $BINDIR"
	if [ -d "$REPO_NAME" -a -d "$REPO_NAME/.git" ] ; then
		echo "Using existing repo: $REPO_NAME"
	else
		echo "Cloning repo from GitHub to $REPO_NAME"
		git clone "$REPO_HOME" "$REPO_NAME"
	fi
	cd "$REPO_NAME"
	git pull
	cd "$OLDPWD"

	cd "$REPO_NAME"
	git checkout develop
	cd "$OLDPWD"
	;;

	esac

	install -v -d -m 0755 "$PREFIX/bin"
	install -v -d -m 0755 "$DOCDIR/hooks"
	for exec_file in $EXEC_FILES ; do
		install -v -m 0755 "$REPO_NAME/$exec_file" "$BINDIR"
	done
	for script_file in $SCRIPT_FILES ; do
		install -v -m 0644 "$REPO_NAME/$script_file" "$BINDIR"
	done
	for hook_file in $HOOK_FILES ; do
		install -v -m 0644 "$hook_file"  "$DOCDIR/hooks"
	done
	exit
	;;
*)
	usage
	exit
	;;
esac


contrib/gitflow-installer.sh
#!/bin/bash

# git-flow make-less installer for *nix systems, by Rick Osborne
# Based on the git-flow core Makefile:
# http://github.com/petervanderdoes/gitflow-avh/blob/master/Makefile

# Licensed under the same restrictions as git-flow:
# http://github.com/CJ-Systems/gitflow-cjs/blob/develop/LICENSE

# Updated for the fork at CJ-Systems

usage() {
	echo "Usage: [environment] igit.sh [install|uninstall] [stable|develop|version] [tag]"
	echo "Environment:"
	echo "   PREFIX=$PREFIX"
	echo "   REPO_HOME=$REPO_HOME"
	echo "   REPO_NAME=$REPO_NAME"
	exit 1
}

# Does this need to be smarter for each host OS?
if [ -z "$PREFIX" ] ; then
	PREFIX="/usr/local"
fi

if [ -z "$REPO_NAME" ] ; then
	REPO_NAME="igit"
fi

if [ -z "$REPO_HOME" ] ; then
	REPO_HOME="https://github.com/ccfiel/gitflow-cjs.git"
fi

EXEC_PREFIX="$PREFIX"
BINDIR="$EXEC_PREFIX/bin"
DATAROOTDIR="$PREFIX/share"
DOCDIR="$DATAROOTDIR/doc/gitflow"

EXEC_FILES="igit"
SCRIPT_FILES="git-flow-init git-flow-feature git-flow-bugfix git-flow-hotfix git-flow-release git-flow-stage git-flow-support git-flow-version gitflow-common gitflow-shFlags git-flow-config"
HOOK_FILES="$REPO_NAME/hooks/*"


echo "### igit no-make installer ###"

case "$1" in
uninstall)
	echo "Uninstalling igit from $PREFIX"
	if [ -d "$BINDIR" ] ; then
		for script_file in $SCRIPT_FILES $EXEC_FILES ; do
			echo "rm -vf $BINDIR/$script_file"
			rm -vf "$BINDIR/$script_file"
		done
		rm -rf "$DOCDIR"
	else
		echo "The '$BINDIR' directory was not found."
	fi
	exit
	;;
help)
	usage
	exit
	;;
install)
	if [ -z $2 ]; then
		usage
		exit
	fi
	echo "Installing igit to $BINDIR"
	if [ -d "$REPO_NAME" -a -d "$REPO_NAME/.git" ] ; then
		echo "Using existing repo: $REPO_NAME"
	else
		echo "Cloning repo from GitHub to $REPO_NAME"
		git clone "$REPO_HOME" "$REPO_NAME"
	fi
	cd "$REPO_NAME"
	git pull
	cd "$OLDPWD"
	case "$2" in
	stable)
		cd "$REPO_NAME"
		git checkout master
		cd "$OLDPWD"
		;;
	develop)
		cd "$REPO_NAME"
		git checkout develop
		cd "$OLDPWD"
		;;
	version)
		cd "$REPO_NAME"
		git checkout tags/$3
		cd "$OLDPWD"
		;;		
	*)
		usage
		exit
		;;
	esac
	install -v -d -m 0755 "$PREFIX/bin"
	install -v -d -m 0755 "$DOCDIR/hooks"
	for exec_file in $EXEC_FILES ; do
		install -v -m 0755 "$REPO_NAME/$exec_file" "$BINDIR"
	done
	for script_file in $SCRIPT_FILES ; do
		install -v -m 0644 "$REPO_NAME/$script_file" "$BINDIR"
	done
	for hook_file in $HOOK_FILES ; do
		install -v -m 0644 "$hook_file"  "$DOCDIR/hooks"
	done
	exit
	;;
*)
	usage
	exit
	;;
esac
